explorer/interpreter/type_checker.cpp

```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
+      : value_(value), type_(type) {}
+
+  // Make a match-anything wildcard pattern.
+  static auto MakeWildcard() -> AbstractPattern {
+    return AbstractPattern(WildcardTag());
+  }
+
+  auto kind() const -> Kind {
+    if (auto* pattern = value_.dyn_cast<const Pattern*>()) {
+      return Compound;
+    }
+    if (auto* value = value_.dyn_cast<const Value*>()) {
+      if (isa<TupleValue, AlternativeValue>(value)) {
+        return Compound;
+      }
+      return Primitive;
+    }
+    assert(value_.is<const WildcardTag*>());
+    return Wildcard;
+  }
+
+  auto discriminator() const -> std::string_view {
+    assert(kind() == Compound);
+    if (auto* pattern = value_.dyn_cast<const Pattern*>()) {
+      if (auto* alt_pattern = dyn_cast<AlternativePattern>(pattern)) {
+        return alt_pattern->alternative_name();
+      }
+    } else if (auto* value = value_.dyn_cast<const Value*>()) {
+      if (auto* alt = dyn_cast<AlternativeValue>(value)) {
+        return alt->alt_name();
+      }
+    }
+    return {};
+  }
+
+  auto NumElements() const -> int {
+    if (auto* pattern = value_.dyn_cast<const Pattern*>()) {
+      if (auto* tuple_pattern = dyn_cast<TuplePattern>(pattern)) {
+        return tuple_pattern->fields().size();
+      } else if (isa<AlternativePattern>(pattern)) {
+        return 1;
+      }
+    } else if (auto* value = value_.dyn_cast<const Value*>()) {
+      if (auto* tuple = dyn_cast<TupleValue>(value)) {
+        return tuple->elements().size();
+      } else if (auto* alt = dyn_cast<AlternativeValue>(value)) {
+        return 1;
+      }
+    }
+    return 0;
+  }
+
+  void AppendElementsTo(std::vector<AbstractPattern>& out) const {
+    if (auto* pattern = value_.dyn_cast<const Pattern*>()) {
+      if (auto* tuple_pattern = dyn_cast<TuplePattern>(pattern)) {
+        auto fields = tuple_pattern->fields();
+        out.insert(out.end(), fields.begin(), fields.end());
+      } else if (auto* alt_pattern = dyn_cast<AlternativePattern>(pattern)) {
+        out.push_back(&alt_pattern->arguments());
+      }
+    } else if (auto* value = value_.dyn_cast<const Value*>()) {
+      if (auto* tuple = dyn_cast<TupleValue>(value)) {
+        auto* tuple_type = cast<TupleValue>(type_);
+        assert(tuple->elements().size() == tuple_type->elements().size());
+        for (size_t i = 0; i != tuple->elements().size(); ++i) {
+          out.push_back(
+              AbstractPattern(tuple->elements()[i], tuple_type->elements()[i]));
+        }
+      } else if (auto* alt = dyn_cast<AlternativeValue>(value)) {
+        out.push_back(AbstractPattern(
+            &alt->argument(),
+            *cast<ChoiceType>(type_)->FindAlternative(alt->alt_name())));
+      }
+    }
+  }
+
+  auto value() const -> const Value& { return *value_.get<const Value*>(); }
+
+  auto type() const -> const Value& {
+    assert(kind() == Compound);
+    return *type_;
+  }
+
+ private:
+  struct alignas(8) WildcardTag {};
```

Comment:
Why does this need `alignas`? (comment?)

Reply:
```suggestion
  // This is aligned so that we can use it in the `PointerUnion` below.
  struct alignas(8) WildcardTag {};
```